libraries/aws/provisioning/src/aws_iot_provisioning_api.c

```diff
@@ -169,51 +202,145 @@ static bool _checkInit( void )
+        if( ( pCallbackInfo->createKeysAndCertificateCallback.function == NULL ) ||
+            ( pCallbackInfo->createCertificateFromCsrCallback.function == NULL ) ||
+            ( pCallbackInfo->registerThingCallback.function == NULL ) )
         {
-            IotLogWarn( "Unknown parsing status on topic %s. Ignoring message.",
-                        pPublishData->u.message.pTopicFilter,
-                        pPublishData->u.message.topicFilterLength );
-            _activeOperation.info.status = AWS_IOT_PROVISIONING_INTERNAL_FAILURE;
+            IotLogDebug( "Received unexpected server response on topic %s.",
+                         pPublishData->u.message.pTopicFilter,
+                         pPublishData->u.message.topicFilterLength );
         }
         else
         {
-            /* Parse the server response and execute the user callback. */
-            _activeOperation.info.status = responseParser(
-                responseStatus,
-                pPublishData->u.message.info.pPayload,
-                pPublishData->u.message.info.payloadLength,
-                &_activeOperation.info.userCallback );
+            /* Determine whether the response from the server is "accepted" or "rejected". */
+            responseStatus = AwsIot_ParseStatus( pPublishData->u.message.pTopicFilter,
+                                                 pPublishData->u.message.topicFilterLength );
+
+            if( responseStatus == AWS_IOT_UNKNOWN )
+            {
+                IotLogWarn( "Unknown parsing status on topic %s. Ignoring message.",
+                            pPublishData->u.message.pTopicFilter,
+                            pPublishData->u.message.topicFilterLength );
+                _activeOperation[ operationIndex ].info.status =
+                    AWS_IOT_PROVISIONING_INTERNAL_FAILURE;
+            }
+            else
+            {
+                /* Parse the server response and execute the user callback. */
+                _activeOperation[ operationIndex ].info.status = responseParser(
+                    responseStatus,
+                    pPublishData->u.message.info.pPayload,
+                    pPublishData->u.message.info.payloadLength,
+                    &_activeOperation[ operationIndex ].info.userCallback );
+            }
+
+            /* Notify the waiting thread about arrival of response from server */
+            /* Increment the number of PUBLISH messages received. */
+            IotSemaphore_Post( &_activeOperation[ operationIndex ].responseReceivedSem );
+
+            /* Invalidate the user-callback information to prevent re-processing the response
+             * if we receive the same response multiple times (which is possible for a QoS 1 publish
+             * from the server). This is done to post on the semaphore ONLY ONCE on receiving the
+             * response from the server.*/
+            memset( &_activeOperation[ operationIndex ].info.userCallback,
```

Comment:
In the QoS1 multiple responses scenarios, can those responses be posted to multiple task pool threads? If so, could this memset be racing with another one of those response-handling task pool threads?

Reply:
Yes, you have rightly pointed at the race condition in the QoS multiple responses case. I will remove this logic.

Reply:
Purged it 🔥